clients/src/main/java/org/apache/kafka/common/security/kerberos/Login.java

```diff
@@ -0,0 +1,410 @@
+            throw new LoginException("loginContext name (JAAS file section header) was null. " +
+                    "Please check your java.security.login.auth.config (=" +
+                    System.getProperty("java.security.login.auth.config") +
+                    ") and your " + JaasUtils.LOGIN_CONTEXT_SERVER + "(=" +
+                    System.getProperty(JaasUtils.LOGIN_CONTEXT_CLIENT, "Client") + ")");
+        }
+
+        if (System.getProperty("java.security.auth.login.config") == null) {
+            throw new IllegalArgumentException("You must pass java.security.auth.login.config in secure mode.");
+        }
+
+        File configFile = new File(System.getProperty("java.security.auth.login.config"));
+        Configuration loginConf = null;
+        try {
+            loginConf = Configuration.getInstance("JavaLoginConfig", new URIParameter(configFile.toURI()));
+            Configuration.setConfiguration(loginConf);
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+
+        LoginContext loginContext = new LoginContext(loginContextName, callbackHandler);
+        loginContext.login();
+        log.info("successfully logged in.");
+        return loginContext;
+    }
+
+    private long getRefreshTime(KerberosTicket tgt) {
+        long start = tgt.getStartTime().getTime();
+        long expires = tgt.getEndTime().getTime();
+        log.info("TGT valid starting at:        " + tgt.getStartTime().toString());
+        log.info("TGT expires:                  " + tgt.getEndTime().toString());
+        long proposedRefresh = start + (long) ((expires - start) *
+                (TICKET_RENEW_WINDOW + (TICKET_RENEW_JITTER * rng.nextDouble())));
+
+        if (proposedRefresh > expires)
+            // proposedRefresh is too far in the future: it's after ticket expires: simply return now.
+            return time.currentWallTime();
+        else
+            return proposedRefresh;
+    }
+
+    private synchronized KerberosTicket getTGT() {
+        Set<KerberosTicket> tickets = subject.getPrivateCredentials(KerberosTicket.class);
+        for (KerberosTicket ticket: tickets) {
+            KerberosPrincipal server = ticket.getServer();
+            if (server.getName().equals("krbtgt/" + server.getRealm() + "@" + server.getRealm())) {
+                log.debug("Found tgt " + ticket + ".");
+                return ticket;
+            }
+        }
+        return null;
+    }
+
+    private boolean hasSufficientTimeElapsed() {
+        long now = time.currentElapsedTime();
+        if (now - getLastLogin() < MIN_TIME_BEFORE_RELOGIN) {
+            log.warn("Not attempting to re-login since the last re-login was " +
+                    "attempted less than " + (MIN_TIME_BEFORE_RELOGIN / 1000) + " seconds" +
+                    " before.");
+            return false;
+        }
+        // register most recent relogin attempt
+        setLastLogin(now);
```

Comment:
That's right, changed it locally. With the current code, this doesn't make much of a difference in practice, but it could lead to bugs in the future.